airbytehq/airbyte
# 🪟 Use source defined primary keys over configured

## What

closes #16098
closes https://github.com/airbytehq/oncall/issues/461

See oncall issue for more context.

Previously, if the primary key changed in a user's source, the UI and platform never actually applied the change.  With this PR, we now consider the source-defined primary key in calculating the `syncCatalog`.

When we refresh a source schema, if the source-defined primary keys have changed, we now override any configured primary keys with what was returned.

Likewise, we evaluate whether the source gives a `sourceDefinedCursor` and overrides the config for that if it changes as well.  

In these contexts, "source defined" means a source has a set primary key and/or cursor field and a user has no way to configure these in the UI.

## How
Adds a method when calculating the initial values for the `syncCatalog` object

## Recommended reading order
Any

## To test:
- create a source that you can change the primary keys of (i ran airbyte in local docker and connected to a local postgres db)
- create a connection with that source and run a sync
- change the primary key of that source and refresh source schema
- see that the primary key is now updated
- click save and confirm that you want to reset
- should still be the updated primary key (this will also be reflected in the sync logs of that reset sync if you really want to dig in)

## 🚨 User Impact 🚨
Users who had previously changed a primary key in a source that is in a connection running `incremental | dedupe` will now have the correct primary keys on their next reset.  Previously, the primary keys would never be overridden by the source-defined ones.
